### PR TITLE
feat(http): merge request body with RequestOptionsArgs body

### LINKED ISSUE #11306: Http post - body param not included in request, when passing  RequestOptionsArgs.body
**I'm submitting a ...**  (check one with "x")

```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

```
this.headers = new Headers(
     { 'Content-Type': 'application/json' }
);
this.options = new RequestOptions({
     headers: this.headers,
     body: {
          default_param: "default_val",
      }
});

action(customBodyParams ): Observable<MyModel> {
        return this.http.post(this.myUrl, customBodyParams, this.options)
            .map(data => data.json().data || {})
            .catch(this.handleError);
    }
```

When I passed third param to post function (RequestOptions) containing body property my customBodyParams have been omitted in request.

**Expected/desired behavior**
I expect to see merged parameters from customBodyParams and this.options.body.
IMHO Request Options should be treated as default request options and shouldn't override value of the second post parameter (body: any)

**Reproduction of the problem**
If the current behavior is a bug or you can illustrate your feature request better with an example, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (you can use this template as a starting point: http://plnkr.co/edit/tpl:AvJOMERrnz94ekVua0u5).

**Please tell us about your environment:**
- **Angular version:** 2.0.0-rc.6
- **Browser:** [all] 
- **Language:** [ TypeScript 2.0 ]

### PR DESCRIPTION
**Pull request by dalamgir**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
https://github.com/angular/angular/issues/11306

**What is the new behavior?**
The body object supplied to the POST, PATCH or PUT methods on the http service is now merged with any default body parameters that may have been specified in a RequestOptionsArgs object also supplied to those methods.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

Merge body parameters with RequestOptionsArgs body parameters if the RequestOptionsArgs param is supplied and has a default body.
The request body is a merged version of the two body parameters. This behavior is available in the POST, PUT and PATCH requests.

Closes #11306
Http post - body param not included in request, when passing RequestOptionsArgs.body

### CODE REVIEW

**Comment by alamgird** [comment]
I signed it!

**Comment by alamgird** [comment]
I authored all of them. How can I change the email address?

**Comment by vicb** [comment]
Please merge the 2 commits or fix the comments.

To change the author of a commit, you can use `git commit --amend` (google for "git change commit author")

**Comment by alamgird** [comment]
@vicb The issue should be fixed now

**Comment by vicb** [comment]
Please squash your commits into a single one and

>  The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format

File: modules/@angular/http/src/http.ts [hunk 1]
```diff
@@ -22,6 +22,22 @@ function httpRequest(backend: ConnectionBackend, request: Request): Observable<R
   return backend.createConnection(request).response;
 }
 
+function objectAssign(out: any, objA: any, objB: any) {
+  out = out || {};
+
+  var objects = [objA, objB];
+
+  for (var i = 0; i < objects.length; i++) {
```
Comment (vicb) [inline modules/@angular/http/src/http.ts]: IMO a recursive call would make more sense here

File: modules/@angular/http/test/http_spec.ts [hunk 2]
```diff
@@ -234,6 +234,23 @@ export function main() {
              });
              http.post(url, body).subscribe((res: Response) => {});
            }));
+
+        it('should merge the request options default body with the body provided to the request',
+           inject([AsyncTestCompleter], (async: AsyncTestCompleter) => {
+             var body = {customParam: 'customParam', anotherCustomParam: 'anotherCustomParam'};
```
Comment (vicb) [inline modules/@angular/http/test/http_spec.ts]: var -> const (feel free to update other tests as well - in a different `refactor` commit)

File: modules/@angular/http/src/http.ts (hunk 1 above)
Comment (alamgird) [inline modules/@angular/http/src/http.ts]: agreed. However this kind of functions ideally should belong in a utils module or the like.

**Comment by alxhub** [comment]
Closing as it is very out of date.

**Pull request closed without merging**